### PR TITLE
http transparent endpoint support,

### PR DESCRIPTION
**Pull request by vitaliy-kuzmich**
hello, I asked before to add http endpoint support, because some of the http proxies failed, so please review changes.

### CODE REVIEW

**Comment by normanmaurer** [comment]
@trustin PTAL

File: handler-proxy/src/main/java/io/netty/handler/proxy/HttpProxyHandler.java [hunk 1]
```diff
@@ -46,6 +46,7 @@
     private final String password;
     private final CharSequence authorization;
     private HttpResponseStatus status;
+    private boolean isTransparent;
```
Comment (trustin) [inline handler-proxy/src/main/java/io/netty/handler/proxy/HttpProxyHandler.java]: `isT` -> `t`

File: handler-proxy/src/main/java/io/netty/handler/proxy/HttpProxyHandler.java [hunk 2]
```diff
@@ -74,6 +75,18 @@ public HttpProxyHandler(SocketAddress proxyAddress, String username, String pass
...
+    /**
+     *
+     * @return endpoint http/ws support
+     */
+    public boolean isTransparent() {
+        return isTransparent;
+    }
+
+    public void setTransparent(boolean transparent) {
+        isTransparent = transparent;
```
Comment (trustin) [inline handler-proxy/src/main/java/io/netty/handler/proxy/HttpProxyHandler.java]: Could we just add a constructor parameter instead of making `transparent` mutable?

File: handler-proxy/src/main/java/io/netty/handler/proxy/HttpProxyHandler.java [hunk 3]
```diff
@@ -158,4 +171,13 @@ protected boolean handleResponse(ChannelHandlerContext ctx, Object response) thr
 
         return finished;
     }
+
+    @Override
+    protected void sendInitialMessage(ChannelHandlerContext ctx) throws Exception {
+        if(isTransparent()){
+            setConnectSuccess();
+        }else {
```
Comment (trustin) [inline handler-proxy/src/main/java/io/netty/handler/proxy/HttpProxyHandler.java]: Please fix the code style.

File: handler-proxy/src/main/java/io/netty/handler/proxy/ProxyHandler.java [hunk 4]
```diff
@@ -280,7 +280,7 @@ public final void channelRead(ChannelHandlerContext ctx, Object msg) throws Exce
      */
     protected abstract boolean handleResponse(ChannelHandlerContext ctx, Object response) throws Exception;
 
-    private void setConnectSuccess() {
+    protected void setConnectSuccess() {
```
Comment (trustin) [inline handler-proxy/src/main/java/io/netty/handler/proxy/ProxyHandler.java]: `protected final`

File: handler-proxy/src/main/java/io/netty/handler/proxy/HttpProxyHandler.java (hunk 2 above)
Comment (normanmaurer) [inline handler-proxy/src/main/java/io/netty/handler/proxy/HttpProxyHandler.java]: +1

**Comment by trustin** [comment]
@vitaliy-kuzmich Left a few comments.

**Comment by vitaliy-kuzmich** [comment]
@trustin what do you mean 'protected final' ? should I move 'transparent' to the ProxyHandler? but then I am not sure about Socks proxy.

code style?

File: handler-proxy/src/main/java/io/netty/handler/proxy/HttpProxyHandler.java [hunk 5]
```diff
@@ -74,6 +67,20 @@ public HttpProxyHandler(SocketAddress proxyAddress, String username, String pass
...
+
+    }
+
+    /**
+     * @return endpoint http/ws support
+     */
+    public boolean isTransparent() {
+        return transparent;
+    }
+
```
Comment (normanmaurer) [inline handler-proxy/src/main/java/io/netty/handler/proxy/HttpProxyHandler.java]: remove empty line

File: handler-proxy/src/main/java/io/netty/handler/proxy/HttpProxyHandler.java [hunk 6]
```diff
@@ -74,6 +67,20 @@ public HttpProxyHandler(SocketAddress proxyAddress, String username, String pass
         authzBase64.release();
     }
 
+    public HttpProxyHandler(SocketAddress proxyAddress, String username, String password, boolean transparent) {
+        this(proxyAddress, username, password);
+        this.transparent = transparent;
+
```
Comment (normanmaurer) [inline handler-proxy/src/main/java/io/netty/handler/proxy/HttpProxyHandler.java]: remove empty line

File: handler-proxy/src/main/java/io/netty/handler/proxy/HttpProxyHandler.java [hunk 7]
```diff
@@ -46,6 +38,7 @@
     private final String password;
     private final CharSequence authorization;
     private HttpResponseStatus status;
+    private boolean transparent;
```
Comment (normanmaurer) [inline handler-proxy/src/main/java/io/netty/handler/proxy/HttpProxyHandler.java]: final (and group with the other final fields).

File: handler-proxy/src/main/java/io/netty/handler/proxy/HttpProxyHandler.java [hunk 8]
```diff
@@ -74,6 +67,20 @@ public HttpProxyHandler(SocketAddress proxyAddress, String username, String pass
...
     }
 
+    public HttpProxyHandler(SocketAddress proxyAddress, String username, String password, boolean transparent) {
+        this(proxyAddress, username, password);
+        this.transparent = transparent;
+
+    }
+
+    /**
+     * @return endpoint http/ws support
```
Comment (normanmaurer) [inline handler-proxy/src/main/java/io/netty/handler/proxy/HttpProxyHandler.java]: Change to `Returns {@code true} if the endpoint has http/ws support transparently` or something similar that makes sense.

File: handler-proxy/src/main/java/io/netty/handler/proxy/ProxyHandler.java [hunk 9]
```diff
@@ -192,7 +192,7 @@ public final void channelActive(ChannelHandlerContext ctx) throws Exception {
      * Sends the initial message to be sent to the proxy server. This method also starts a timeout task which marks
      * the {@link #connectPromise} as failure if the connection attempt does not success within the timeout.
      */
-    private void sendInitialMessage(final ChannelHandlerContext ctx) throws Exception {
+    protected void sendInitialMessage(final ChannelHandlerContext ctx) throws Exception {
```
Comment (normanmaurer) [inline handler-proxy/src/main/java/io/netty/handler/proxy/ProxyHandler.java]: make package private (remove protected)

File: handler-proxy/src/main/java/io/netty/handler/proxy/HttpProxyHandler.java [hunk 10]
```diff
@@ -81,7 +88,7 @@ public String protocol() {
 
     @Override
     public String authScheme() {
-        return authorization != null? AUTH_BASIC : AUTH_NONE;
+        return authorization != null ? AUTH_BASIC : AUTH_NONE;
```
Comment (normanmaurer) [inline handler-proxy/src/main/java/io/netty/handler/proxy/HttpProxyHandler.java]: revert this change

File: handler-proxy/src/main/java/io/netty/handler/proxy/HttpProxyHandler.java [hunk 11]
```diff
@@ -21,15 +21,7 @@
...
-import io.netty.handler.codec.http.DefaultFullHttpRequest;
-import io.netty.handler.codec.http.FullHttpRequest;
-import io.netty.handler.codec.http.HttpClientCodec;
-import io.netty.handler.codec.http.HttpHeaderNames;
-import io.netty.handler.codec.http.HttpMethod;
-import io.netty.handler.codec.http.HttpResponse;
-import io.netty.handler.codec.http.HttpResponseStatus;
-import io.netty.handler.codec.http.HttpVersion;
-import io.netty.handler.codec.http.LastHttpContent;
+import io.netty.handler.codec.http.*;
```
Comment (normanmaurer) [inline handler-proxy/src/main/java/io/netty/handler/proxy/HttpProxyHandler.java]: please revert this and so not use wildcard import.

**Comment by normanmaurer** [comment]
@vitaliy-kuzmich please address my comments and then squash the commits + adjust the commit message to match our template:
http://netty.io/wiki/writing-a-commit-message.html

File: handler-proxy/src/main/java/io/netty/handler/proxy/HttpProxyHandler.java [hunk 12]
```diff
@@ -52,9 +53,10 @@ public HttpProxyHandler(SocketAddress proxyAddress) {
         username = null;
         password = null;
         authorization = null;
+        transparent = false;
     }
 
-    public HttpProxyHandler(SocketAddress proxyAddress, String username, String password) {
+    public HttpProxyHandler(SocketAddress proxyAddress, String username, String password, boolean transparent) {
```
Comment (trustin) [inline handler-proxy/src/main/java/io/netty/handler/proxy/HttpProxyHandler.java]: Could you keep this constructor and add a new one with `transparent` parameter?

**Comment by vitaliy-kuzmich** [comment]
As you know your code better, please fix HttpProxyServer, read "HTTP tunneling without using CONNECT" for more details, https://en.wikipedia.org/wiki/HTTP_tunnel.

**Comment by trustin** [comment]
I'm not sure I understood the intention of this PR. What do you want to achieve with this? Netty's HttpProxyHandler is meant to use `CONNECT` and nothing else.

**Comment by vitaliy-kuzmich** [comment]
hmm okay then, just cut commit then.

**Comment by vitaliy-kuzmich** [comment]
@trustin how can I cut this commit?
